### PR TITLE
`Integrated code lifecycle`: Improve SSH handling

### PR DESCRIPTION
**Pull request by SimonEntholzer**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types locally.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

Users should use SSH to clone from localVC for programming exercise repositories.

### Description

- Added settings subpage to add, edit and remove the users SSH key
- Added link in the copy clone link UI to the SSH key settings
- Added documentation about SSH for new users

Link to SSH settings in the clone dialogue:
image
SSH settings page: - creating a new key
image
save the key:
image
edit or delete the key:
image

### Steps for Testing

Prerequisites:

1. Log in to Artemis as a student, go to Settings
2. Test if you can add, edit and remove your SSH key, in the SSH Settings tab.
3. Go to a programming exercise, copy the clone link, and check if you can clone the exercise with the key you set in the settings.

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

## Summary by CodeRabbit

- **New Features**
  - Added functionality to manage SSH public keys within user settings, including viewing, editing, and deleting keys.
  - Introduced new constants and configuration for GitLab version control support.

- **Bug Fixes**
  - Updated conditions in the clone repository button component to correctly display warnings related to SSH key usage.

- **Documentation**
  - Added new links for SSH setup in both English and German documentation.

- **Tests**
  - Introduced tests for the SSH User Settings component, ensuring proper initialization, key saving, and edit mode toggling.

### CODE REVIEW

File: src/main/webapp/app/core/auth/account.service.ts [hunk 1]
```diff
@@ -333,4 +333,11 @@ export class AccountService implements IAccountService {
         }
         return this.http.put<void>('api/users/sshpublickey', sshPublicKey);
     }
+
+    deleteSshPublicKey() {
+        if (this.userIdentity) {
+            this.userIdentity.sshPublicKey = '';
```
Comment (Strohgelaender) [inline src/main/webapp/app/core/auth/account.service.ts:339]: Why an empty string?
```suggestion
            this.userIdentity.sshPublicKey = undefined;
```

File: src/main/webapp/app/shared/user-settings/ssh-settings/ssh-user-settings.component.html [hunk 2]
```diff
@@ -0,0 +1,81 @@
+<h1>
+    <!--Account Information-->
+    {{ 'artemisApp.userSettings.sshSettings' | artemisTranslate }}
+</h1>
+@if (currentUser) {
+    <div class="list-group d-block">
+        <div class="list-group-item">
+            <dt>
+                {{ 'artemisApp.userSettings.sshSettingsPage.sshSettingsInfoText' | artemisTranslate }}
+                <a href="https://docs.artemis.cit.tum.de/user/exercises/programming.html#repository-access">
```
Comment (Strohgelaender) [inline src/main/webapp/app/shared/user-settings/ssh-settings/ssh-user-settings.component.html:10]: Could you use the existing documentation button component for this?

**Review by BBesrour — COMMENTED** [review]
Tested on TS2:
- Can edit and delete ssh key
- warning is visible and link redirects me to the settings

Should the warning still be visible even if the user has set an ssh key?
And I am not sure if this is related to your PR but I get a 500 error when I access any course
image
image

**Comment by Strohgelaender** [comment]
>Should the warning still be visible even if the user has set an ssh key?

Imo it would be better if the warning would disapear now, see also #6650. Before we were not able to know if the user has a ssh key setup, now we can test for that and should hide the message.

**Comment by SimonEntholzer** [comment]
> > Should the warning still be visible even if the user has set an ssh key?
> 
> Imo it would be better if the warning would disapear now, see also #6650. Before we were not able to know if the user has a ssh key setup, now we can test for that and should hide the message.

Good point, I changed it so that the warning goes away when there is a SSH key available. :+1:

File: src/main/webapp/app/core/auth/account.service.ts (hunk 1 above)
Comment (SimonEntholzer) [reply src/main/webapp/app/core/auth/account.service.ts]: personal taste I guess

File: src/main/webapp/app/shared/user-settings/ssh-settings/ssh-user-settings.component.html (hunk 2 above)
Comment (SimonEntholzer) [reply src/main/webapp/app/shared/user-settings/ssh-settings/ssh-user-settings.component.html:10]: Ah that's a good tip, didn't know it existed :+1: I'll add it

File: src/main/webapp/app/shared/user-settings/ssh-settings/ssh-user-settings.component.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    readonly documentationType: DocumentationType = 'SshSetup';
+    currentUser?: User;
+    localVCEnabled = false;
+    sshKey = '';
+    storedSshKey = '';
+    editSshKey = false;
+
+    faEdit = faEdit;
+    faSave = faSave;
+    faTrash = faTrash;
```
Comment (JohannesStoehr) [inline src/main/webapp/app/shared/user-settings/ssh-settings/ssh-user-settings.component.ts:25]:
```suggestion
    readonly faEdit = faEdit;
    readonly faSave = faSave;
    readonly faTrash = faTrash;
```

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]

File: src/main/webapp/app/shared/components/clone-repo-button/clone-repo-button.component.html [hunk 4]
```diff
@@ -14,7 +14,7 @@
         container="body"
     ></button>
     <ng-template #popContent>
-        @if (useSsh && !localVCEnabled) {
+        @if ((useSsh && !user.sshPublicKey) || gitlabVCEnabled) {
```
[the remark this answers is not in the document]
Comment (SimonEntholzer) [reply src/main/webapp/app/shared/components/clone-repo-button/clone-repo-button.component.html:17]: That's at least the current behaviour, which I wanted to keep. As artemis can't find out if the user has a SSH key set in gitlab, I'd keep it that way for now. Another option, as mentioned in https://github.com/ls1intum/Artemis/issues/6650 would be to add a close button, and store that the notification was closed

File: src/main/webapp/app/shared/user-settings/ssh-settings/ssh-user-settings.component.ts (hunk 3 above)
Comment (SimonEntholzer) [reply src/main/webapp/app/shared/user-settings/ssh-settings/ssh-user-settings.component.ts:25]: good point :+1:

**Review by beyzaaltuntas — CHANGES_REQUESTED** [review]

File: src/main/webapp/app/shared/user-settings/ssh-settings/ssh-user-settings.component.html [hunk 5]
```diff
@@ -0,0 +1,79 @@
+<h1>
+    <!--Account Information-->
+    {{ 'artemisApp.userSettings.sshSettings' | artemisTranslate }}
```
Comment (sarpsahinalp) [inline src/main/webapp/app/shared/user-settings/ssh-settings/ssh-user-settings.component.html:3]: Consider using jhiTranslate

File: src/main/webapp/app/shared/user-settings/ssh-settings/ssh-user-settings.component.html [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+</h1>
+@if (currentUser) {
+    <div class="list-group d-block">
+        <div class="list-group-item">
+            <dt>
+                {{ 'artemisApp.userSettings.sshSettingsPage.sshSettingsInfoText' | artemisTranslate }}
```
Comment (sarpsahinalp) [inline src/main/webapp/app/shared/user-settings/ssh-settings/ssh-user-settings.component.html:9]: Also `jhiTranslate`

File: src/main/webapp/app/shared/user-settings/ssh-settings/ssh-user-settings.component.html [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+                <jhi-documentation-button [type]="documentationType" />
+            </dt>
+        </div>
+        @if (storedSshKey === '' && !editSshKey) {
+            <div class="list-group-item">
+                <button class="btn btn-sm btn-info" (click)="editSshKey = !editSshKey">
+                    <fa-icon [icon]="faEdit" />
+                    {{ 'artemisApp.userSettings.sshSettingsPage.addNewSshKey' | artemisTranslate }}
```
Comment (sarpsahinalp) [inline src/main/webapp/app/shared/user-settings/ssh-settings/ssh-user-settings.component.html:17]: See above

**Review by sarpsahinalp — COMMENTED** [review]
Consider using `jhiTranslate `instead of `{{ 'property' | artemisTranslate }}`

File: src/main/webapp/app/shared/user-settings/ssh-settings/ssh-user-settings.component.html [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+                </button>
+            </div>
+        }
```
[the remark this answers is not in the document]
Comment (SimonEntholzer) [reply src/main/webapp/app/shared/user-settings/ssh-settings/ssh-user-settings.component.html]: Now the alertService shows a notificatoin when a key has been successfully saved :+1:

File: src/main/webapp/app/shared/components/clone-repo-button/clone-repo-button.component.ts [hunk 9]
```diff
@@ -67,14 +69,21 @@ export class CloneRepoButtonComponent implements OnInit, OnChanges {
...
                 this.versionControlUrl = profileInfo.versionControlUrl;
             }
+
             this.versionControlAccessTokenRequired = profileInfo.versionControlAccessToken;
             this.localVCEnabled = profileInfo.activeProfiles.includes(PROFILE_LOCALVC);
+            this.gitlabVCEnabled = profileInfo.activeProfiles.includes(PROFILE_GITLAB);
+
+            if (this.localVCEnabled) {
+                this.setupSshKeysUrl = `${window.location.origin}/user-settings/sshSettings`;
+            }
```
[the remark this answers is not in the document]
Comment (SimonEntholzer) [reply src/main/webapp/app/shared/components/clone-repo-button/clone-repo-button.component.ts]: I added this fallback :+1:

File: src/main/java/de/tum/in/www1/artemis/web/rest/UserResource.java [hunk 10]
```diff
@@ -193,4 +194,19 @@ public ResponseEntity<Void> addSshPublicKey(@RequestBody String sshPublicKey) th
...
+
+    /**
+     * PUT users/sshpublickey : sets the ssh public key
+     *
+     * @return the ResponseEntity with status 200 (OK), with status 404 (Not Found), or with status 400 (Bad Request)
+     */
+    @DeleteMapping("users/sshpublickey")
+    @EnforceAtLeastStudent
+    public ResponseEntity<Void> deleteSshPublicKey() {
+        User user = userRepository.getUser();
```
[the remark this answers is not in the document]
Comment (SimonEntholzer) [reply src/main/java/de/tum/in/www1/artemis/web/rest/UserResource.java:219]: good idea, I added it :+1:

File: src/main/webapp/app/shared/components/clone-repo-button/clone-repo-button.component.html (hunk 4 above)
Comment (Strohgelaender) [inline src/main/webapp/app/shared/components/clone-repo-button/clone-repo-button.component.html:17]: Only show the warning if the user wants to use SSH.
```suggestion
        @if (useSsh && (!user.sshPublicKey || gitlabVCEnabled)) {
```

Comment (SimonEntholzer) [reply src/main/webapp/app/shared/components/clone-repo-button/clone-repo-button.component.html:17]: good suggestion :+1:

File: src/main/java/de/tum/in/www1/artemis/web/rest/UserResource.java [hunk 11]
```diff
@@ -184,13 +190,37 @@ public ResponseEntity<Void> setIrisAcceptedToTimestamp() {
...
+        log.debug("REST request to add SSH key to user {}", user.getLogin());
         // Parse the public key string
-        AuthorizedKeyEntry keyEntry = AuthorizedKeyEntry.parseAuthorizedKeyEntry(sshPublicKey);
+        AuthorizedKeyEntry keyEntry;
+        try {
+            keyEntry = AuthorizedKeyEntry.parseAuthorizedKeyEntry(sshPublicKey);
+        }
+        catch (IllegalArgumentException e) {
+            return ResponseEntity.badRequest().headers(HeaderUtil.createFailureAlert(applicationName, true, "sshUserSettings", "saveSshKeyError", e.getMessage())).body(null);
+        }
```
[the remark this answers is not in the document]
Comment (SimonEntholzer) [reply src/main/java/de/tum/in/www1/artemis/web/rest/UserResource.java]: good suggestion

**Review by BBesrour — APPROVED** [review]
tested on TS1, lgtm!

**Review by DominikRemo — APPROVED** [review]
Tested in testing session on ts1. LGTM

**Review by EneaGore — APPROVED** [review]
Demonestrated in testing session in TS1. Lgtm

**Review by coolchock — APPROVED** [review]
tested on ts1 during testing session, LGTM

**Review by kaancayli — APPROVED** [review]
code

**Review by janthoXO — COMMENTED** [review]
i would change the color of the here. As the color of the normal text is already white, the white of here does not communicate a link

**Review by janthoXO — COMMENTED** [review]
the redirect (here, info button in settings) are not consistent in their behavior:
The information button in the ssh settings redirects the user to the documentation by not opening a new tab but just a normal redirect.
On the other hand the "To use ssh, you need to add an ssh key to your account here" link redirects by opening a new tab

**Comment by SimonEntholzer** [comment]
> i would change the color of the here. As the color of the normal text is already white, the white of here does not communicate a link

I'm not sure what you mean. For me the "here" is blue in light mode and orange in dark mode, and therefore clearly identifiable as a link. Might be an issue with the browser you are using.

**Comment by janthoXO** [comment]
> > i would change the color of the here. As the color of the normal text is already white, the white of here does not communicate a link
> 
> I'm not sure what you mean. For me the "here" is blue in light mode and orange in dark mode, and therefore clearly identifiable as a link. Might be an issue with the browser you are using.

that's what it looks like for me on Arc Browser
image

**Review by janthoXO — APPROVED** [review]

**Review by kaancayli — APPROVED** [review]
Reapprove

**Review by EneaGore — APPROVED** [review]
re-approve

**Review by BBesrour — APPROVED** [review]
re-approve

**Review by Strohgelaender — APPROVED** [review]
code

**Pull request merged**